storage/innobase/handler/ha_innodb.cc

```diff
@@ -10118,6 +10118,11 @@ wsrep_append_key(
 )
 {
 	DBUG_ENTER("wsrep_append_key");
+	if (trx->is_bulk_insert())
+	{
+		DBUG_RETURN(0);
+	}
+
```

Comment:
The location to bail out from individual row key append feels to happen unnecessary late. Note that the actual key value is already generated at this point (wsrep_append_key), and this function is called for all DML (insert, update,delete). I suggest to move the bulk insert check in the end of ha_innobase::write_row(), where wsrep_append_rows() is called. There is a bunch of other exceptions, as well, to skip wsrep_append_rows() call.
In this location, a debug level assert could be planted, to verify that bulk insert never leaks into this code path
i.e.: ut_ad(!trx->is_bulk_insert())

Reply:
Fixed.